pages/design/test-cases.md

```diff
@@ -0,0 +1,68 @@
+
+- inapplicable: visible, but **not** included in the accessibility tree
+- inapplicable: **not** visible, but included in the accessibility tree
+- inapplicable: **not** visible, and **not** included in the accessibility tree
+
+## Consistency With Accessibility Requirements
+
+In order for the mapping of implementations to the rules to work, test cases need to be consistent with the accessibility requirement. This requires that **passed and inapplicable test cases should satisfy the accessibility requirements** of the rule, as well as the accessibility requirement of any composite rule this rule is designed for. Inapplicable conditions that do not satisfy the accessibility requirements should not be included.
+
+Related to this, failed test cases should only fail the accessibility requirements for one reason. A failed test case for 4.1.2 Name, Role, Value should not have issues with both the accessible name, and the role. If the rule is about accessible names, the role and value must be correct.
+
+## Small Test Cases
+
+A general principle in test case creation is to keep test cases small. Avoid any unnecessary content, as it may cause unexpected side effects in the outcome. For example, a rule about buttons does not require that all test cases have an HTML element with a `lang` attribute and a `title` element. Those elements do not affect the accessibility of a button.
+
+An exception to that, is that ACT Rules should strive to show good accessibility practices. A rule about valid `autocomplete` attributes should still provide labels for the form controls with the `autocomplete` attribute. Those labels both provide context, showing the intended purpose of the input, and they show good practice of providing labels for form controls.
+
+## Superficial Definition Testing
+
+ACT rules use definition to avoid duplication, and to "hide" away the details. Definitions such as "semantic role", "visible" and "accessible name" are used in many rules, and are fundamental building blocks of ACT. To avoid having a lot of similar test cases in every rule that uses a particular definition, it is best to keep definition testing to a minimum.
+
+As an example, the "X has non-empty accessible name" rules all have test cases that check for the different ways that element can have an accessible name. For `img`, that is `alt`, `title`, and `aria-label(ledby)` attributes. That creates sufficient coverage to ensure the basics are understood. But those rules do not go into the details of ensuring accessible name computation is done right. Whether an implementor considers `aria-owns` in the accessible name computation is not relevant for "non-empty accessible name" rules.
+
+That is not to say edge cases in definition should never have test cases. If there are **real-world examples** of a particular edge case affecting the outcome of a rule, test cases should be added to the rule to cover that edge case.
```

Comment:
Random thinking of probably bad solutions for these problems, hoping that bad ideas can be picked up by others and evolve into good ones…

---
A way to deep test shared definitions without duplicating corner cases in every rules would be to have a fake rule dedicated to that purpose. For example, a rule "element has accessible name" without any accessibility requirements, just for the sake of piling up all the corner cases of accessible name computation.
Problem with that is creating a stupid rule that can be ignored by implementors, and the fact that accessible name computation is not an exact science (as in every single combination of UA+AT solves corner cases differently…)

---
A way to know which definition is of which kind could be to have some "usage guideline" info inside the definition. Either as a label in the frontmatter, or as a somewhat lengthier definition. Thinks like "when using this definition, write test cases for each of the conditions in it" (for something like "hidden state"), or "this definition is thoroughly tested in rule Foo, not need to retest everything in other rules" (for something like background colors), or …
Problem is that these may evolve, require some initial work to be set up, and may require a fair amount of work to maintain.

Reply:
Definitions, by definition, are a "formal statement of the meaning or significance or a word, phrase" etc. For ACT Rules, essentially explaining an agreed way to understand a word or phrase in respect of the ACT Rules. They aren't really hiding detail, and I don't think they should be. They help ACT Rules to remain more succinct through using the defined word or phrase, rather than explaining it inline each time.

I don't think definitions need tests. However, if something about a definition has implications for a particular rule that may require additional test cases, I do think that should be handled within each rule where that is applicable. Those cases are relevant for the rule, not the definition.